cmd/fpga_admissionwebhook/fpga_admissionwebhook.go

```diff
@@ -86,7 +152,98 @@ func getTLSConfig(certFile string, keyFile string) *tls.Config {
 	}
 }
 
-func mutatePods(ar v1beta1.AdmissionReview) *v1beta1.AdmissionResponse {
+func getPatchOpsPreprogrammed(containerIdx int, container corev1.Container) ([]string, error) {
+	var ops []string
+
+	for resourceName, resourceQuantity := range container.Resources.Limits {
+		newName := translateFpgaResourceName(resourceName)
+		if len(newName) > 0 {
+			op := fmt.Sprintf(resourceReplaceOp, containerIdx,
+				"limits", rfc6901Escaper.Replace(string(resourceName)),
+				containerIdx, "limits", newName, resourceQuantity.String())
+			ops = append(ops, op)
+		}
+	}
+	for resourceName, resourceQuantity := range container.Resources.Requests {
+		newName := translateFpgaResourceName(resourceName)
+		if len(newName) > 0 {
+			op := fmt.Sprintf(resourceReplaceOp, containerIdx,
+				"requests", rfc6901Escaper.Replace(string(resourceName)),
+				containerIdx, "requests", newName, resourceQuantity.String())
+			ops = append(ops, op)
+		}
+	}
+
+	return ops, nil
+}
+
+func getEnvVars(container corev1.Container) string {
+	var jsonstrings []string
+	for _, envvar := range container.Env {
+		if envvar.Name == "FPGA_REGION" || envvar.Name == "FPGA_AFU" {
+			continue
+		}
+		jsonbytes, err := json.Marshal(envvar)
+		if err != nil {
+			glog.Warning("Skipping '%s' because of %v", envvar.Name, err)
+			continue
+		}
+		jsonstrings = append(jsonstrings, string(jsonbytes))
+	}
+
+	if len(jsonstrings) == 0 {
+		return ""
+	}
+
+	return fmt.Sprintf(", %s", strings.Join(jsonstrings, ","))
+}
+
+func getPatchOpsOrchestrated(containerIdx int, container corev1.Container) ([]string, error) {
+	var ops []string
+	var mutated bool
+
+	for resourceName, resourceQuantity := range container.Resources.Limits {
+		interfaceID, afuID, ok := parseResourceName(string(resourceName))
+		if !ok {
+			continue
+		}
+
+		if mutated {
+			return nil, fmt.Errorf("Only one FPGA resource per container is supported in '%s' mode", orchestrated)
+		}
+
+		op := fmt.Sprintf(resourceReplaceOp, containerIdx, "limits", rfc6901Escaper.Replace(string(resourceName)),
+			containerIdx, "limits", rfc6901Escaper.Replace("intel.com/fpga-region-"+interfaceID), resourceQuantity.String())
+		ops = append(ops, op)
+
+		op = fmt.Sprintf(envAddOp, containerIdx, interfaceID, afuID, getEnvVars(container))
+		ops = append(ops, op)
+		mutated = true
+	}
+
+	mutated = false
+	for resourceName, resourceQuantity := range container.Resources.Requests {
+		interfaceID, _, ok := parseResourceName(string(resourceName))
+		if !ok {
+			continue
+		}
+
+		if mutated {
+			return nil, fmt.Errorf("Only one FPGA resource per container is supported in '%s' mode", orchestrated)
+		}
+
+		op := fmt.Sprintf(resourceReplaceOp, containerIdx, "requests", rfc6901Escaper.Replace(string(resourceName)),
+			containerIdx, "requests", rfc6901Escaper.Replace("intel.com/fpga-region-"+interfaceID), resourceQuantity.String())
+		ops = append(ops, op)
+		mutated = true
+	}
+
```

Comment:
2 above for loops look almost identical. Can we reorganize the code to avoid it?

Reply:
They look similar, but are not identical. I don't know how to reorganize them without sacrificing readability.